### DISCUSSION TITLE
Erro ao testar os métodos que exigem autenticação

### DISCUSSION THREAD

**Discussion by AleDinizRJ** [balta-io/7194#13] [Perguntas e Respostas] [answered]
Pessoal, bom dia!
Ao testar todos os métodos, pelo Postman, que exigem alguma autenticação, eu sempre recebo um erro 401 Unauthorized. Mesmo passando o Token no Headers. Eu não sei se o problema é no token gerado. Eu vou postar o código aqui e a imagem do erro:

![image](https://user-images.githubusercontent.com/88723782/128877554-beceaa65-b1a7-485f-b314-8b5f76af1d4c.png)

**Código do Settings.cs:**
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Shop
{
    public static class Settings
    {
        public static string Secret = "aleRJ8239testalfac##balta8866";
    }
}
-----------------------------------------------------------------------------------------------------------------

**Código do TokenService.cs:**

using Microsoft.IdentityModel.Tokens;
using Shop.Models;
using System;
using System.Collections.Generic;
using System.IdentityModel.Tokens.Jwt;
using System.Linq;
using System.Security.Claims;
using System.Text;
using System.Threading.Tasks;

namespace Shop.Services
{
    public static class TokenService
    {
        public static string GenerateToken(User user)
        {
            var tokenHandler = new JwtSecurityTokenHandler();
            var key = Encoding.ASCII.GetBytes(Settings.Secret);
            var tokenDescriptor = new SecurityTokenDescriptor
            {
                Subject = new ClaimsIdentity(new Claim[]
                {
                    new Claim(ClaimTypes.Name, user.UserName.ToString()),
                    new Claim(ClaimTypes.Role, user.Role.ToString())
                }),
                Expires = DateTime.UtcNow.AddHours(2),
                SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
            };
            var token = tokenHandler.CreateToken(tokenDescriptor);
            return tokenHandler.WriteToken(token);
        }
    }
}
-----------------------------------------------------------------------------------------------------------------------------

**Código do Login do UserController:**
 [HttpPost]
        [Route("login")]
        [AllowAnonymous]
        public async Task<ActionResult<dynamic>> Authenticate(
                    [FromServices] DataContext context,
                    [FromBody] User model)
        {
            var user = await context.Users
                .AsNoTracking()
                .Where(x => x.UserName == model.UserName && x.Password == model.Password)
                .FirstOrDefaultAsync();

            if (user == null)
                return NotFound(new { message = "Usuário ou senha inválidos" });

            var token = TokenService.GenerateToken(user);
            // Esconde a senha
            user.Password = "";
            return new
            {
                user = user,
                token = token
            };
        }

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia @AleDinizRJ, chegou a comparar os códigos com os do curso? Pode ser que tenha errado algo na digitação. Em adicional, tenta rodar o código do curso e ver se o erro contece também.

Verifica estes itens e retorna aqui!

**Comment by AleDinizRJ** [comment]
Eu já copiei o código do curso responsável pela geração do token. Porém não
consegui resolver ainda.

Em qui., 12 de ago. de 2021 às 05:28, Andre Baltieri <
***@***.***> escreveu:

> Bom dia @AleDinizRJ <https://github.com/AleDinizRJ>, chegou a comparar os
> códigos com os do curso? Pode ser que tenha errado algo na digitação. Em
> adicional, tenta rodar o código do curso e ver se o erro contece também.
>
> Verifica estes itens e retorna aqui!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balta-io/7194/discussions/13#discussioncomment-1162461>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVE5CRVOIZWPKNKAAWVB2NLT4OAZDANCNFSM5B4GVY6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Atc,

Alessandro Diniz

**Reply by andrebaltieri** [reply]
Bom dia @AleDinizRJ, rodei o código do curso e está tudo OK! Também não tivemos outro report sobre um erro parecido.
Vou te pedir que revise o código para ver onde errou!

**Comment by AleDinizRJ** [comment]
Ok. Balta. Vou fazer isso. Obrigado!

Em seg., 16 de ago. de 2021 às 10:05, Andre Baltieri <
***@***.***> escreveu:

> Bom dia @AleDinizRJ <https://github.com/AleDinizRJ>, rodei o código do
> curso e está tudo OK! Também não tivemos outro report sobre um erro
> parecido.
> Vou te pedir que revise o código para ver onde errou!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balta-io/7194/discussions/13#discussioncomment-1189877>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVE5CRXHLLLH4ALHSX4O2NDT5EEH3ANCNFSM5B4GVY6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Atc,

Alessandro Diniz
